src/detect-tls-validity.c

```diff
@@ -0,0 +1,1008 @@
+    sigmatch_table[DETECT_AL_TLS_NOTBEFORE].RegisterTests = TlsNotBeforeRegisterTests;
+
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].name = "tls_notafter";
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].desc = "match TLS certificate notAfter field";
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].url = "https://redmine.openinfosecfoundation.org/projects/suricata/wiki/TLS-keywords#tlsnotafter";
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].Match = NULL;
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].AppLayerMatch = DetectTlsValidityMatch;
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].alproto = ALPROTO_TLS;
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].Setup = DetectTlsNotAfterSetup;
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].Free = DetectTlsValidityFree;
+    sigmatch_table[DETECT_AL_TLS_NOTAFTER].RegisterTests = TlsNotAfterRegisterTests;
+
+    const char *eb;
+    int eo;
+    int opts = 0;
+
+    parse_regex = pcre_compile(PARSE_REGEX, opts, &eb, &eo, NULL);
+    if (parse_regex == NULL) {
+        SCLogError(SC_ERR_PCRE_COMPILE,"pcre compile of \"%s\" failed at offset %"
+                   PRId32 ": %s", PARSE_REGEX, eo, eb);
+        goto error;
+    }
+
+    parse_regex_study = pcre_study(parse_regex, 0, &eb);
+    if (eb != NULL) {
+        SCLogError(SC_ERR_PCRE_STUDY,"pcre study failed: %s", eb);
+        goto error;
+    }
+
+error:
+    return;
+}
+
+/**
+ * \internal
+ * \brief Function to match validity field in a tls certificate.
+ *
+ * \param t       Pointer to thread vars.
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param s       Pointer to the Signature.
+ * \param m       Pointer to the sigmatch that we will cast into
+ *                DetectTlsValidityData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectTlsValidityMatch (ThreadVars *t, DetectEngineThreadCtx *det_ctx,
+                   Flow *f, uint8_t flags, void *state, Signature *s, SigMatch *m)
+{
+    SCEnter();
+
+    SSLState *ssl_state = (SSLState *)state;
+    if (ssl_state == NULL) {
+        SCLogDebug("no tls state, no match");
+        SCReturnInt(0);
+    }
+
+    int ret = 0;
+
+    SSLStateConnp *connp = NULL;
+    if (flags & STREAM_TOSERVER)
+        connp = &ssl_state->client_connp;
+    else
+        connp = &ssl_state->server_connp;
+
+    const DetectTlsValidityData *dd = (const DetectTlsValidityData *)m->ctx;
+
+    time_t cert_epoch = 0;
+    if (dd->type == DETECT_TLS_TYPE_NOTBEFORE)
+        cert_epoch = connp->cert0_not_before;
+    else if (dd->type == DETECT_TLS_TYPE_NOTAFTER)
+        cert_epoch = connp->cert0_not_after;
+
+    if (cert_epoch == 0)
+        SCReturnInt(0);
+
+    if (dd->mode == DETECT_TLS_VALIDITY_EQ && dd->epoch == cert_epoch)
+        ret = 1;
+    else if (dd->mode == DETECT_TLS_VALIDITY_LT && dd->epoch > cert_epoch)
```

Comment:
_LT followed by > is a bit confusing. Also, it seems that by using 'keyword epoch > cert epoch', it really is 'cert epoch <= keyword epoch'. So LEQ instead of LT.